### PR TITLE
【Gemma3】add gemma3(text)

### PR DESCRIPTION
**Pull request by lijialin03**
#### Before submitting

- [x] Lint code. If there are lint issues, please format the code first.

```shell
# Install and register `pre-commit` in the project folder
pip install pre-commit && pre-commit install

# Process previous code files separately
pre-commit run --file XXXX.py
```

- [x] Add test cases into `tests` folder. If there are codecov issues, please add tests cases first.

### PR types

Others
### PR changes

Models
### Description

**支持：**
- Gemma3中text部分的组网
- 相关单测
- 动态图推理
- model.generate()生成
- PP
- SFT(Full & LoRA)
- DPO (Full & LoRA)
- PT (Full & LoRA)
- flexcheckpoint

**不支持：**
- TP（有相关适配但没有验证条件）
- SP（有部分适配但未验证正确性）
- EP（非MOE未适配）
- 根据 layer_type 设置 RoPE

**future：**
- 星河社区(AIStuido)模型下载：上传模型&增加映射

### CODE REVIEW

File: paddleformers/transformers/gemma3_text/configuration.py [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+        "layers.*.self_attn.o_proj": "rowwise",
+        "layers.*.mlp.gate_proj": "colwise",
+        "layers.*.mlp.up_proj": "colwise",
+        "layers.*.mlp.down_proj": "rowwise",
+    }
+    base_model_pp_plan = {
+        "embed_tokens": (["input_ids"], ["inputs_embeds"]),
+        "layers": (["hidden_states", "attention_mask"], ["hidden_states"]),
+        "norm": (["hidden_states"], ["hidden_states"]),
+    }
```
Comment (WYB27) [inline paddleformers/transformers/gemma3_text/configuration.py]: base_model_tp/pp_plan paddleformers里面不需要写，可以删了

File: paddleformers/transformers/gemma3_text/configuration.py [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+        self.sliding_window = sliding_window
+        self.final_logit_softcapping = final_logit_softcapping
+        self.attn_logit_softcapping = attn_logit_softcapping
+        self.layer_types = layer_types
+        self.use_bidirectional_attention = use_bidirectional_attention
+        if use_bidirectional_attention:
+            self.sliding_window = self.sliding_window // 2 + 1
+        self.rope_local_base_freq = rope_local_base_freq
+        self.rope_scaling = rope_scaling
+        # rope_config_validation(self)
```
Comment (WYB27) [inline paddleformers/transformers/gemma3_text/configuration.py]: 没用到就删了吧，包括导入

File: paddleformers/transformers/gemma3_text/configuration.py [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+            text_config = Gemma3TextConfig()
+            logger.info("text_config is None, using default Gemma3TextConfig text config.")
+        elif isinstance(text_config, dict):
+            text_config = Gemma3TextConfig(**text_config)
+
+        # if isinstance(vision_config, dict):
+        #     vision_config = SiglipVisionConfig(**vision_config)
+        # elif vision_config is None:
+        #     vision_config = SiglipVisionConfig()
+        #     logger.info("vision_config is None, using default SiglipVisionConfig vision config.")
```
Comment (WYB27) [inline paddleformers/transformers/gemma3_text/configuration.py]: vision_config的内容是为后续接vision模型保留吗？

File: paddleformers/transformers/gemma3_text/configuration.py [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+        image_token_index (`int`, *optional*, defaults to 262144):
+            The image token index to encode the image prompt.
+        initializer_range (`float`, *optional*, defaults to 0.02):
+            The standard deviation of the truncated_normal_initializer for initializing all weight matrices.
+
+
+    Example:
+
+    ```python
+    >>> from transformers import Gemma3ForConditionalGeneration, Gemma3Config, SiglipVisionConfig, Gemma3TextConfig
```
Comment (WYB27) [inline paddleformers/transformers/gemma3_text/configuration.py]: 有关transformers的注释/示例 都对应改成paddleformers

File: paddleformers/transformers/gemma3_text/modeling.py [hunk 5]
```diff
@@ -0,0 +1,931 @@
...
+        else:
+            self.rope_type = "default"
+        self.max_seq_len_cached = config.max_position_embeddings
+        self.original_max_seq_len = config.max_position_embeddings
+
+        self.config = config
+        # TODO: rope_type is default in `gemma-3-1b-it` but linear in `gemma-3-4b-it`
+        self.rope_init_fn = _compute_default_rope_parameters
+
+        inv_freq, self.attention_scaling = self.rope_init_fn(self.config, device)
```
Comment (WYB27) [inline paddleformers/transformers/gemma3_text/modeling.py]: 计算inv_freq如果没有特殊要求，能否简化一点？可以参考https://github.com/PaddlePaddle/PaddleFormers/blob/c9de6d11c0d56c3286b1b565d04ae229ffda4900/paddleformers/transformers/glm4_moe/modeling.py#L1064

File: paddleformers/transformers/gemma3_text/modeling.py [hunk 6]
```diff
@@ -0,0 +1,931 @@
...
+            max_sequence_length = self.config.max_sequence_length
+            bsz = hidden_states.shape[0] * self.config.tensor_parallel_degree // max_sequence_length
+            q_len = max_sequence_length
+        else:
+            bsz, q_len, _ = hidden_states.shape
+
+        hidden_shape = (bsz, q_len, -1, self.head_dim)
+
+        query_states = self.q_proj(hidden_states).reshape(hidden_shape).transpose(1, 2)
+        key_states = self.k_proj(hidden_states).reshape(hidden_shape).transpose(1, 2)
```
Comment (WYB27) [inline paddleformers/transformers/gemma3_text/modeling.py]: 建议与其他模型保持一致，query_states, key_states先不要在外部transpose(1, 2)，保持原有形状，进入apply_rotary_pos_emb再进行transpose，可以参考
https://github.com/PaddlePaddle/PaddleFormers/blob/c9de6d11c0d56c3286b1b565d04ae229ffda4900/paddleformers/transformers/qwen3/modeling.py#L184

File: paddleformers/transformers/gemma3_text/modeling.py [hunk 7]
```diff
@@ -0,0 +1,931 @@
...
+        # if we have enough num_key_value_heads to split, then split it.
+        if config.num_key_value_heads % config.tensor_parallel_degree == 0:
+            LAYER_COLWISE.extend(["self_attn.k_proj.weight", "self_attn.v_proj.weight"])
+            BIAS_KEYS.extend(["self_attn.k_proj.bias", "self_attn.v_proj.bias"])
+
+        def make_base_actions():
+            if config.vocab_size % config.tensor_parallel_degree == 0:
+                actions = {
+                    "lm_head.weight": partial(fn, is_column=False),
+                    "embed_tokens.weight": partial(fn, is_column=False),
```
Comment (WYB27) [inline paddleformers/transformers/gemma3_text/modeling.py:454]: 需要确认一下模型权重里还有没有别的以embed_tokens.weight结尾的参数，避免踩前后覆盖的坑

File: paddleformers/transformers/gemma3_text/modeling.py [hunk 8]
```diff
@@ -0,0 +1,931 @@
...
+        # See https://github.com/huggingface/transformers/pull/29402
+        self.embed_tokens = Gemma3TextScaledWordEmbedding(config)
+        self.layers = nn.LayerList(
+            [Gemma3DecoderLayer(config, layer_idx) for layer_idx in range(config.num_hidden_layers)]
+        )
+        self.norm = Gemma3RMSNormPipe(config)
+        self.rotary_emb = Gemma3RotaryEmbedding(config=config)
+
+        # TODO: raushan fix this after RoPE refactor. For now we hack it by reassigning thetas
+        # when we want to create a local RoPE layer. Config defaults should hold values for global RoPE
```
Comment (WYB27) [inline paddleformers/transformers/gemma3_text/modeling.py]: 所有不相关的注释辛苦清理一下

File: paddleformers/transformers/gemma3_text/modeling.py [hunk 9]
```diff
@@ -0,0 +1,931 @@
...
+        config = copy.deepcopy(config)
+        config.rope_theta = config.rope_local_base_freq
+        config.rope_scaling = {"rope_type": "default"}
+        self.rotary_emb_local = Gemma3RotaryEmbedding(config=config)
```
Comment (WYB27) [inline paddleformers/transformers/gemma3_text/modeling.py]: 这里不像huggingface一样在Gemma3RotaryEmbedding里通过config区分不同emb的原因是？这样下面position_embeddings可以统一调用，减少和其他组网的区别

File: paddleformers/transformers/gemma3_text/modeling.py [hunk 10]
```diff
@@ -0,0 +1,931 @@
...
+            output_hidden_states=output_hidden_states,
+            attn_mask_startend_row_indices=attn_mask_startend_row_indices,
+            return_dict=return_dict,
+            **kwargs,
+        )
+
+        hidden_states = outputs[0]
+
+        # Only compute necessary logits, and do not upcast them to float if we are not computing the loss
+        slice_indices = slice(-logits_to_keep, None) if isinstance(logits_to_keep, int) else logits_to_keep
```
Comment (WYB27) [inline paddleformers/transformers/gemma3_text/modeling.py]: logits_to_keep的逻辑，paddleformers不需要，可以删了

File: paddleformers/transformers/gemma3_text/modeling.py (hunk 5 above)
Comment (lijialin03) [reply paddleformers/transformers/gemma3_text/modeling.py]: 这里是因为像上面TODO说的，gemma3其他大小的模型用到的reop_init_fn类型不同，在原代码中它是通过`self.rope_init_fn = ROPE_INIT_FUNCTIONS[self.rope_type]`来指定的，所以这里就采用了类似的写法，如果直接写死的话，就是相当于当前仅支持rope_type=default时的情况

File: paddleformers/transformers/gemma3_text/modeling.py (hunk 9 above)
Comment (lijialin03) [reply paddleformers/transformers/gemma3_text/modeling.py]: 在复现这段时间内huggingface的代码进行了一些改动，包括这个位置，会参考最新代码整体检查一下

File: paddleformers/transformers/gemma3_text/modeling.py (hunk 5 above)
Comment (lijialin03) [reply paddleformers/transformers/gemma3_text/modeling.py]: 已修改，当前只支持`default `类的ROPE

File: paddleformers/transformers/gemma3_text/modeling.py (hunk 6 above)
Comment (lijialin03) [reply paddleformers/transformers/gemma3_text/modeling.py]: 已修改

File: paddleformers/transformers/gemma3_text/modeling.py (hunk 7 above)
Comment (lijialin03) [reply paddleformers/transformers/gemma3_text/modeling.py:454]: 已确认没有

File: paddleformers/transformers/gemma3_text/modeling.py (hunk 8 above)
Comment (lijialin03) [reply paddleformers/transformers/gemma3_text/modeling.py]: 参考QWen3，留下了所有todo和部分代码注释

File: paddleformers/transformers/gemma3_text/modeling.py (hunk 9 above)
Comment (lijialin03) [reply paddleformers/transformers/gemma3_text/modeling.py]: 由于套件中rope utils相关部分还在建设，当前模型代码基于8.14版本实现

**Review by lugimzzz — APPROVED** [review]
LGTM
